atomics/include/desul/atomics/Fetch_Op_SYCL.hpp

```diff
@@ -11,52 +11,125 @@ SPDX-License-Identifier: (BSD-3-Clause)
 
 #include <desul/atomics/Adapt_SYCL.hpp>
 #include <desul/atomics/Common.hpp>
+#include <desul/atomics/Operator_Function_Objects.hpp>
 
 namespace desul {
 namespace Impl {
 
 // clang-format off
-#define DESUL_IMPL_SYCL_ATOMIC_FETCH_OPER(OPER, TYPE)                                     \
-  template <class MemoryOrder>                                                            \
-  TYPE device_atomic_fetch_##OPER(TYPE* dest, TYPE val, MemoryOrder, MemoryScopeDevice) { \
-    sycl_atomic_ref<TYPE, MemoryOrder, MemoryScopeDevice> dest_ref(*dest);                \
-    return dest_ref.fetch_##OPER(val);                                                    \
-  }                                                                                       \
-  template <class MemoryOrder>                                                            \
-  TYPE device_atomic_fetch_##OPER(TYPE* dest, TYPE val, MemoryOrder, MemoryScopeCore  ) { \
-    sycl_atomic_ref<TYPE, MemoryOrder, MemoryScopeCore> dest_ref(*dest);                  \
-    return dest_ref.fetch_##OPER(val);                                                    \
+#define DESUL_IMPL_SYCL_ATOMIC_OPER(OPER, TYPE)                                             \
```

Comment:
The code duplication for MemoryScopeCore and MemoryScopeDevice is increased 8x.
Did you consider strategies to avoid that blowing up?

Reply:
I prioritized improving the performance without obfuscating too much what really was changed. I was thinking about trying to minimize the lines of code in a follow-up.

Reply:
If you make your macro take a trailing scope argument, you don't obfuscate anything and you drop many lines of code